src/combobox/ComboBox.ts

```diff
@@ -0,0 +1,389 @@
+ * @property onMenuChange		Called when menu visibility changes
+ */
+export interface ComboBoxProperties extends ThemeableProperties {
+	autoBlur?: boolean;
+	clearable?: boolean;
+	customResultItem?: any;
+	customResultMenu?: any;
+	disabled?: boolean;
+	formId?: string;
+	inputProperties?: any;
+	invalid?: boolean;
+	label?: string | LabelOptions;
+	openOnFocus?: boolean;
+	readOnly?: boolean;
+	required?: boolean;
+	results?: any[];
+	value?: string;
+	getResultLabel?(result: any): string;
+	onBlur?(value: string): void;
+	onChange?(value: string): void;
+	onFocus?(value: string): void;
+	onRequestResults?(value: string): void;
+	onMenuChange?(open: boolean): void;
+};
+
+const enum Operation {
+	increase,
+	decrease,
+	reset
+};
+
+// This can go away when Safari decides to support KeyboardEvent.key
+const keys = {
+	escape: 27,
+	enter: 13,
+	up: 38,
+	down: 40
+};
+
+export const ComboBoxBase = ThemeableMixin(WidgetBase);
+
+@theme(css)
+export default class ComboBox extends ComboBoxBase<ComboBoxProperties> {
+	private _ignoreBlur: boolean;
+	private _ignoreFocus: boolean;
+	private _inputElement: HTMLInputElement;
+	private _focused: boolean;
+	private _open: boolean;
+	private _direction: Operation;
+	private _selectedIndex: number | undefined;
+	private _wasOpen: boolean;
+
+	private _handlers: {[key: string]: any} = {
+		[keys.up](this: ComboBox, event: KeyboardEvent) {
+			event.preventDefault();
+			this.updateSelectedIndex(Operation.decrease);
+			this.invalidate();
+		},
+
+		[keys.down](this: ComboBox, event: KeyboardEvent) {
+			event.preventDefault();
+			this.updateSelectedIndex(Operation.increase);
+			this.invalidate();
+		},
+
+		[keys.enter](this: ComboBox) {
+			const { results } = this.properties;
+
+			// If results already closed or no results, do nothing
+			if (!this._open || !results || results.length === 0) {
+				return;
+			}
+			// If no result selected, close menu
+			else if (this._selectedIndex === undefined) {
+				this._open = false;
+				this.invalidate();
+			}
+			else {
+				this.selectResult(results[this._selectedIndex]);
+			}
+		},
+
+		[keys.escape](this: ComboBox) {
+			this.updateSelectedIndex(Operation.reset);
+			this._open = false;
+			this.invalidate();
+		}
+	};
+
+	updateSelectedIndex(operation?: Operation, index?: number) {
+		const { results } = this.properties;
+
+		if (!results || results.length === 0 || !this._open) {
+			return;
+		}
+
+		if (index !== undefined) {
+			this._selectedIndex = index;
+		}
+		else if (operation !== undefined) {
+			if (operation === Operation.increase || operation === Operation.decrease) {
+				const increment = operation === Operation.increase ? 1 : -1;
+				const total = results.length;
+				const current = this._selectedIndex;
+				const base = operation === Operation.increase ? -1 : 0;
+
+				this._selectedIndex = ((current === undefined ? base : current) + increment + total) % total;
```

Comment:
Yes. Seems convoluted, but it's correct. I think your idea to add values to the enum will make this cleaner.